pkg/auth/auth.go

```diff
@@ -0,0 +1,306 @@
+}
+
+func CallLoginEndpoint() (string, error) {
+	res, err := http.Get(authUrlBase + "/login")
+	if err != nil {
+		return "", err
+	}
+	defer closenicely.OrDebug(res.Body)
+	if res.StatusCode < 200 || res.StatusCode >= 300 {
+		return "", errors.Errorf(`received %v from auth server`, res.StatusCode)
+	}
+	body, err := io.ReadAll(res.Body)
+	if err != nil {
+		return "", err
+	}
+	result := LoginResponse{}
+	err = json.Unmarshal(body, &result)
+	if err != nil {
+		return "", err
+	}
+	err = browser.OpenURL(result.Url)
+	if err != nil {
+		return "", err
+	}
+	return result.State, nil
+}
+
+func CallGetTokenEndpoint(state string) error {
+	values := map[string]string{"state": state}
+	jsonData, err := json.Marshal(values)
+	if err != nil {
+		log.Fatal(err)
+	}
+	res, err := http.Post(authUrlBase+"/logintoken", "application/json", bytes.NewBuffer(jsonData))
+	if err != nil {
+		return err
+	}
+	defer closenicely.OrDebug(res.Body)
+	if res.StatusCode != 200 {
+		return fmt.Errorf("recieved invalid status code %d", res.StatusCode)
+	}
+	body, err := io.ReadAll(res.Body)
+	if err != nil {
+		return err
+	}
+	err = WriteIDToken(string(body))
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func CallLogoutEndpoint() error {
+	res, err := http.Get(authUrlBase + "/logout")
+	if err != nil {
+		return errors.Wrap(err, "couldn't invoke logout URL")
+	}
+	defer closenicely.OrDebug(res.Body)
+	body, err := io.ReadAll(res.Body)
+	if err != nil {
+		return errors.Wrap(err, "couldn't read logout redirect URL")
+	}
+	err = browser.OpenURL(string(body))
+	if err != nil {
+		zap.S().Debug("couldn't open logout URL: %s", string(body))
+		zap.L().Warn("couldn't open logout URL. If this persists, run with --verbose to see it. Will still clear local credentials.")
+	}
+
+	configPath, err := cli_config.KlothoConfigPath("credentials.json")
+	if err != nil {
+		return err
+	}
+	if _, err := os.Stat(configPath); err == nil {
+		err = os.Remove(configPath)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func CallRefreshToken(token string) error {
+	values := map[string]string{"refresh_token": token}
+	jsonData, err := json.Marshal(values)
+	if err != nil {
+		return err
+	}
+	res, err := http.Post(authUrlBase+"/refresh", "application/json", bytes.NewBuffer(jsonData))
+	if err != nil {
+		return err
+	}
+	defer closenicely.OrDebug(res.Body)
+	body, err := io.ReadAll(res.Body)
+	if err != nil {
+		return err
+	}
+	err = WriteIDToken(string(body))
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+type KlothoClaims struct {
+	ProEnabled    bool
+	ProTier       int
+	Email         string `json:"email"`
+	EmailVerified bool   `json:"email_verified"`
+	Name          string `json:"name"`
+	jwt.RegisteredClaims
+}
+
+func Authorize() (*KlothoClaims, error) {
+	return authorize(false)
+}
+
+func authorize(tokenRefreshed bool) (*KlothoClaims, error) {
+	creds, claims, err := getClaims()
+	if err != nil {
+		return nil, err
+	}
+
+	if !claims.EmailVerified {
```

Comment:
As written, this means that if the claim has an unverified email, and _also_ is expired and/or not enabled for pro, then we'll allow it (albeit with a warning). Should we move this `if` to the end of the if-else chain, so that the expiration and pro-check take precedence?

Reply:
Talked w/ Jordon offline:

1. This is only a problem while we fail-open
2. So, rather than change the order, I'm going to have the other branches _also_ return both the claims and the err.